sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -36,6 +37,13 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
       .union(constructIsNotNullConstraints(constraints))
       .filter(constraint =>
         constraint.references.nonEmpty && constraint.references.subsetOf(outputSet))
+      .map(_.transform {
```

Comment:
Instead of adding this here, how about we add something like `.map(_.transform { case IsNotNull(Casts(a)) => IsNotNull(a) })` in `constructIsNotNullConstraints`? Wouldn't that avoid the redundant `outputSet` check?